csrc/host_ir/container.h

```diff
@@ -0,0 +1,49 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <fusion.h>
+
+namespace nvfuser {
+
+namespace hir {
+
+/*
+HostIrContainer is used to represent a host program.
+1) It inherits from Fusion, so that (Host) IRs can be resgistered to it.
+2) It holds a vector of Host Expressions `top_level_exprs_` that represent the
+host program. For now, this vector is manually managed. Moreover, because we use
+a vector as data structure, top_level_exprs_ can only represent linear Host
+programs. Later, we it should support non-linear program having a DAG structure.
+*/
+
+class HostIrContainer final : public Fusion {
+ public:
+  HostIrContainer() = default;
+  HostIrContainer(const HostIrContainer&) = delete;
+  HostIrContainer& operator=(const HostIrContainer&) = delete;
+
+  //! Print to an output stream
+  std::ostream& print(std::ostream& os) const;
+
+  const auto& topLevelExprs() const {
+    return top_level_exprs_;
+  }
+
+  void pushBackTopLevelExprs(Expr* expr) {
+    assertInContainer(expr, "Cannot add expr, ");
+    return top_level_exprs_.push_back(expr);
+  }
+
+ private:
+  std::vector<Expr*> top_level_exprs_;
```

Comment:
Given IrContainer [holds](https://github.com/NVIDIA/Fuser/blob/5af71046e8993adcb0a36eb35aaa74d864d10df8/csrc/ir/container.h#L120) a list of Exprs, I failed to see why another Expr list is needed.

Reply:
Given the current design of `IrContainer`, having another list manually managed in `HostIrContainer` is helpful because 
1) `unordered_exprs` returns the list of *all* the registered Exprs, not the ones that constitute the Host program top level instruction. Imagine for example a program of the form
```
ForLoop (i=0; i<N;i++) {
    Expr1;
}
```
then `unordered_exprs` will contain `{ForLoop, Expr1}` while `topLevelExpr` will only contain `{ForLoop}`.
Even with this PR's first example in the test, the host program looks like
```
tv1 = PostOnStream(HostUnit1)(tv0)
``` 
in this case, because `HostUnit1` is an `Expr`, `unordered_exprs` will contain `{PostOnStream, HostUnit1}` while `topLevelExpr` will only contain `{PostOnStream}`
2) `unordered_exprs` is unordered so we cannot rely on it. We want to be able to manually manage the order of the instructions.
3) It helps separate registering the Expr (which is mainly used to manage the pointer's lifetime IIUC) and the actual building of the HostProgram using those Exprs.


Not that the same is done in `kir::Kernel`, see [here](https://github.com/NVIDIA/Fuser/blob/5af71046e8993adcb0a36eb35aaa74d864d10df8/csrc/kernel.h#L205)